internal/controller/compute/user/user.go

```diff
@@ -0,0 +1,208 @@
+func Setup(mgr ctrl.Manager, l logging.Logger, _ workqueue.RateLimiter, opts *utils.ConfigurationOptions) error {
+	name := managed.ControllerName(v1alpha1.UserGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewController(),
+		}).
+		For(&v1alpha1.User{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.UserGroupVersionKind),
+			managed.WithExternalConnecter(&connectorUser{
+				kube:  mgr.GetClient(),
+				usage: resource.NewProviderConfigUsageTracker(mgr.GetClient(), &apisv1alpha1.ProviderConfigUsage{}),
+				log:   l,
+			}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithInitializers(),
+			managed.WithPollInterval(opts.GetPollInterval()),
+			managed.WithTimeout(opts.GetTimeout()),
+			managed.WithCreationGracePeriod(opts.GetCreationGracePeriod()),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+// A connectorUser is expected to produce an ExternalClient when its Connect method
+// is called.
+type connectorUser struct {
+	kube  client.Client
+	usage resource.Tracker
+	log   logging.Logger
+}
+
+// Connect typically produces an ExternalClient by:
+// 1. Tracking that the managed resource is using a ProviderConfig.
+// 2. Getting the managed resource's ProviderConfig.
+// 3. Getting the credentials specified by the ProviderConfig.
+// 4. Using the credentials to form a client.
+func (c *connectorUser) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	svc, err := clients.ConnectForCRD(ctx, mg, c.kube, c.usage)
+	return &externalUser{
+		service: &userapi.APIClient{IonosServices: svc},
+		log:     c.log,
+	}, err
+}
+
+// externalUser observes, then either creates, updates, or deletes a
+// user in the cloud api. it ensures it reflects a desired state.
+type externalUser struct {
+	// service is the ionos cloud api client to manage users.
+	// see https://api.ionos.com/docs/cloud/v6/#tag/User-management
+	service userapi.Client
+	log     logging.Logger
+}
+
+func (eu *externalUser) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.User)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotUser)
+	}
+
+	userID := meta.GetExternalName(cr)
+	if userID == "" {
+		return managed.ExternalObservation{}, nil
+	}
+
+	observed, resp, err := eu.service.GetUser(ctx, userID)
+	if resp.HttpNotFound() {
+		return managed.ExternalObservation{}, nil
+	}
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, "failed to get user by id")
+	}
+
+	props := observed.GetProperties()
+	cr.Status.AtProvider.UserID = *observed.GetId()
+	cr.Status.AtProvider.S3CanonicalUserID = *props.GetS3CanonicalUserId()
```

Comment:
possible nil pointer

Reply:
It will help more if you can provide a concrete example or test case. Lines above check for err and the sdk will never return a `nil` User with a `nil` `err` at the same time. https://github.com/ionos-cloud/sdk-go/blob/master/api_user_management.go#L616-L746

Reply:
From the code it looks like props.GetS3CanonicalUserId() returns a pointer. if that is true, then this pointer can be nil, thus we need to check its value before using it.  
This is the method signature: func (o *UserProperties) GetS3CanonicalUserId() *string. This is similar to my other nil pointer comments.

Reply:
`S3CanonicalUserId` is computed by the ionos api. if the create request succeeds, the value will exists.

Reply:
This is not what the contract says. It says it can be nil(func (o *UserProperties) GetS3CanonicalUserId() _*string_) if what you are saying is correct the contract will be func (o *UserProperties) GetS3CanonicalUserId() _string_

In all cases, this is a violation of defensive programming https://deviq.com/practices/defensive-programming